text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+    }
+}
+
+impl<T> InlineSingleStore<T> {
+    fn validate_layout(layout: Layout) -> Result<(), AllocError> {
+        let own = Layout::new::<T>();
+
+        if layout.align() <= own.align() && layout.size() <= own.size() {
+            Ok(())
+        } else {
+            Err(AllocError)
+        }
+    }
+}
+```
+
+And that's it!
+
+I need not implement `MultipleStore`, and thus do not have to track allocations and deallocations. And I need not
+implement `PinningStore`, and thus do not have to ensure memory address stability across moves.
+
+More examples of `Store` can be found at https://github.com/matthieu-m/storage/tree/main/src/store, including an inline
+bump store.
+
+
+#   Reference-level explanation
+
+##  Overview
+
+This RFC introduces 4 new traits.
+
+The core of this RFC is the `Store` trait:
+
+```rust
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store {
+    /// Handle to a block of memory.
+    type Handle: Copy;
+
+    /// Returns a dangling handle, always invalid.
+    fn dangling() -> Self::Handle;
+
+    /// Return a pointer to the block of memory associated to `handle`.
+    unsafe fn resolve(&self, handle: Self::Handle) -> NonNull<u8>;
+
+    //  The following methods are similar to `Allocator`, reformulated in terms of `Self::Handle`.
+
+    /// Allocates a new handle to a block of memory. On success, invalidates any existing handle.
+    fn allocate(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Deallocates a handle.
+    unsafe fn deallocate(&self, handle: Self::Handle, layout: Layout);
+
+    /// Grows the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn grow(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Shrinks the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn shrink(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Allocates a new handle to a block of zeroed memory. On success, invalidates any existing handle.
+    fn allocate_zeroed(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+
+    /// Grows the block of memory associated to a handle with zeroed memory. On success, the handle is invalidated.
+    unsafe fn grow_zeroed(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+}
+```
+
+_Note:  full-featured documentation of the trait and methods can be found in the companion repository at_
+        https://github.com/matthieu-m/store/blob/main/src/interface.rs.
```

Comment:
Is documentation supposed to be part of the RFC?

I'm happy to have a detailed section documenting the effects/pre-conditions/post-conditions/safety aspects etc... of each method, but I feel like copy/pasting the documentation is maybe not the best way to achieve this.

Reply:
The RFC should fully explain the functionality of any added APIs, and that's the purpose of the API documentation. If examples are present in the RFC prose, not duplicating them is reasonable.

The important part is that all of the information should be available in the RFC. The fact that the API documentation is designed to provide that information in a useful manner makes it a very useful way of presenting that information.

Reply:
I have a serious doubt API documentation is the best format. Or I am doing it wrong.

If you look at the documentation, you'll notice a hefty amount of repetition, because each method tends to fully enunciate the safety requirements. This is easier for the user, I find, as they don't need to hunt down and piece together all the bits and pieces when looking at one particular method. It also works better with IDEs, which tend to only show the one current method documentation.

But if we're talking about a holistic approach to the API, the safety requirements can be expressed by groups of methods, or at the type level, in a much more succinct manner.

Reply:
Note that users of the `Store` API in the future aren't going to have the RFC text available; they're just going to see the API docs and need to learn from there how to use the API. That's why I say having the API text is useful; it's not particularly good at the *why* (that's what the RFC text is explaining), but it does need to be good at the *what* and *how*. The API docs are in fact put in an awkward spot of being primarily reference documentation but needing to satisfy some amount of exploratory explanation as well, for e.g. when someone sees `Box<T, S: Store>` and clicks on `Store` to figure out what that parameter is utilized for.

Reply:
> Note that users of the `Store` API in the future aren't going to have the RFC text available

Obviously, but I still don't see that as a good argument for including the documentation in the RFC itself.

For now, I'm of the opinion that the documentation would do more harm than good in the RFC, as it's _precisely_ formatted for user of the APIs, and not for reviewers of the RFC.

I also don't think it's necessary to review the documentation in the RFC itself, and would rather postpone the discussion to the implementation PR, _after_ the API and its behavior has been settled. At the moment, any word smithing on the documentation seems pointless, given the API and behavior are in flux.